### ISSUE TITLE
Connection pooling

### ISSUE THREAD

**Issue by gregbair** [flamencist/ldap4net#51]
**Is your feature request related to a problem? Please describe.**
Opening new ldap connections is very expensive. It is better to reuse them, but this requires passing connections around your codebase. This is not an ideal solution.

**Describe the solution you'd like**
Many data access libraries contain a connection pooling mechanism, like Microsoft.Data.SqlClient for C#, and [ldaptive](http://www.ldaptive.org/javadocs/org/ldaptive/pool/package-summary.html) for Java. I'd like a connection pooling mechanism for ldap4net.

**Describe alternatives you've considered**
Passing connections around.

**Comment by flamencist** [comment]
I agree connection pool is great feature. I will appreciate if you help to implement the feature.

**Comment by gregbair** [comment]
We've done some work around this at my job. I'll see if we can clean it up and contribute it.

**Comment by flamencist** [comment]
Will be good.

**Comment by nhontran** [comment]
Hi @gregbair , could you share how you implemented the connection pool?

**Comment by FinHorsley** [comment]
Jumping on this late, but is LdapConnection thread safe? I can't see any documentation that explicitly states that it is

Update: just read https://github.com/flamencist/ldap4net/issues/138, and yes it is 🙂
